include/rmm/cuda_stream.hpp

```diff
@@ -0,0 +1,132 @@
+/*
+ * Copyright (c) 2020, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <rmm/cuda_stream_view.hpp>
+#include <rmm/detail/error.hpp>
+
+#include <cuda_runtime_api.h>
+
+#include <memory>
+
+namespace rmm {
+
+/**
+ * @brief Owning wrapper for a CUDA stream.
+ *
+ * Provides RAII lifetime semantics for a CUDA stream.
+ */
+class cuda_stream {
+ public:
+  /**
+   * @brief Move constructor (default)
+   *
+   * A moved-from cuda_stream is invalid and it is Undefined Behavior to call methods that access
+   * the owned stream.
+   */
+  cuda_stream(cuda_stream&&) = default;
+  /**
+   * @brief Move copy assignment operator (default)
+   *
+   * A moved-from cuda_stream is invalid and it is Undefined Behavior to call methods that access
+   * the owned stream.
+   */
+  cuda_stream& operator=(cuda_stream&&) = default;
+  ~cuda_stream()                        = default;
+  cuda_stream(cuda_stream const&)       = delete;  // Copying disallowed: one stream one owner
+  cuda_stream& operator=(cuda_stream&) = delete;
+
+  /**
+   * @brief Construct a new cuda stream object
+   *
+   * @throw rmm::cuda_error if stream creation fails
+   */
+  cuda_stream()
```

Comment:
Do we want to worry about allowing the stream to be non-blocking or specifying a priority at this time?

Reply:
If you don't mind I think those features are sufficiently orthogonal to do them later.

Reply:
Makes sense to me. I was thinking the same.